### PR TITLE
Verify watermark

### PR DESCRIPTION
**Pull request by asim29**
I made several changes to the watermarking module to make it consistent with the rest of the API and refactored one of the functions to act as a verification function. 

However, the only way to "verify" the model is by setting a threshold for the watermark accuracy. However, the original paper does not define a particular threshold. Moreover, the original paper does not test this against model stealing attacks, only fine-tuned and pruned models. Not sure if this would work. I've arbitrarily set this threshold to 90% after some rudimentary testing.

### CODE REVIEW

**Comment by asim29** [comment]
Also, the trigger set is stored in the `data` folder in the project's root directory, which git ignores. If we want to give a default option, it might be worthwhile to set up a link where people can download it OR figure out a directory in the repo where it makes sense to keep it. It's only 100 images and about 6MB.

File: amulet/unauth_model_ownership/defenses/watermark.py [hunk 1]
```diff
@@ -91,52 +80,70 @@ def get_wm_loader(self, shape: int, wm_path: str, gray: bool, tabular: bool):
...
-            wm_targets = np.loadtxt(os.path.join(wm_path, "labels-cifar.txt"))
-            for idx, (path, target) in enumerate(wmset.imgs):
-                img_nlbl.append((path, int(wm_targets[idx])))
-            wmset.imgs = img_nlbl
+            if gray:
+                transform_wm = transforms.Compose(
+                    [
+                        transforms.CenterCrop((shape, shape)),
+                        transforms.Grayscale(num_output_channels=1),
+                        # transforms.ToTensor(),
```
Comment (sebszyller) [inline amulet/unauth_model_ownership/defenses/watermark.py:88]: left over comments

File: amulet/unauth_model_ownership/defenses/watermark.py [hunk 2]
```diff
@@ -91,52 +80,70 @@ def get_wm_loader(self, shape: int, wm_path: str, gray: bool, tabular: bool):
...
-                wmset,
+                trigger_set,
                 batch_size=self.batch_size,
                 shuffle=True,
-                num_workers=4,
-                pin_memory=True,
             )
         return wm_loader
 
     def watermark(self):
```
Comment (sebszyller) [inline amulet/unauth_model_ownership/defenses/watermark.py:112]: missing return type

File: amulet/unauth_model_ownership/defenses/watermark.py [hunk 3]
```diff
@@ -51,35 +27,48 @@ class WatermarkNN:
...
+
+        if isinstance(wm_path, str):
+            wm_path = Path(wm_path)
+
         self.wm_loader = self.get_wm_loader(shape, wm_path, gray, tabular)
-        self.wmtest()
+        self.verify(self.target_model)
 
-    def get_wm_loader(self, shape: int, wm_path: str, gray: bool, tabular: bool):
+    def get_wm_loader(self, shape: int, wm_path: Path, gray: bool, tabular: bool):
```
Comment (sebszyller) [inline amulet/unauth_model_ownership/defenses/watermark.py:71]: missing return type

File: amulet/unauth_model_ownership/defenses/watermark.py [hunk 4]
```diff
@@ -147,154 +154,41 @@ def watermark(self):
...
-                    model.train()
+                    self.verify(self.target_model)
+                    self.target_model.train()
         print("Finished Watermark Embeddding")
-        return model
+        return self.target_model
 
-    def wmtest(self):
-        self.model.eval()
+    def verify(self, model: nn.Module, threshold: int = 90):
```
Comment (sebszyller) [inline amulet/unauth_model_ownership/defenses/watermark.py:162]: return type.

Also, it might be better to have a float threshold, and assume that in this specific case, the accuracy is `0.0-1.0`, to cover things like non-accuracy-based verification methods, e.g. some reconstruction distance for generative models, or p-value based.

File: miscellaneous/trigger_set/images/1.jpg
Comment (sebszyller) [inline miscellaneous/trigger_set/images/1.jpg:1]: commenting only here but applies to all images. We'd include some source info somewhere. What's the license on these images? Can we provide exact urls?

**Review by sebszyller — CHANGES_REQUESTED** [review]

File: amulet/datasets/__data.py [hunk 5]
```diff
@@ -37,3 +43,46 @@ class AmuletDataset:
...
+
+        self.img_labels = pd.read_csv(labels_file)
+
+        if self.img_labels.shape[1] != 2:
+            raise Exception("Labels file should have 2 columns.")
+
+        self.img_dir = img_dir
+        self.transform = transform
+
+    def __len__(self):
```
Comment (sebszyller) [inline amulet/datasets/__data.py:78]: `-> int:`

File: miscellaneous/trigger_set/images/1.jpg
Comment (asim29) [reply miscellaneous/trigger_set/images/1.jpg:1]: I believe it's an MIT license. This is the source URL: https://github.com/adiyoss/WatermarkNN/tree/master/data/trigger_set

**Comment by asim29** [comment]
This should be good to go now. I've replaced the trigger set with 100 random images from the ImageNet test set, and updated the README.

**Review by sebszyller — APPROVED** [review]
